### DISCUSSION TITLE
State synchronization across server nodes.

### DISCUSSION THREAD

**Discussion by bwasia** [colyseus/colyseus#578] [Q&A] [answered]
Does Colyseus synchronize state across servers? Suppose if there's two servers Server A and Server B running on ports 2567 and 2568 respectively. They both share the same server codebase.

Please refer to the code snippets below. `index.ts` is the server code, followed by console outputs from Server A and B.

Server A and B connects to a Redis docker container running on port `6379`. As shown in the console outputs, the state is not shared.

index.ts:
```typescript
import "dotenv/config.js";
const port = process.env.COLYSEUS_PORT
    ? Number(process.env.COLYSEUS_PORT)
    : 2567;
import { RedisPresence, Room, Server } from "colyseus";
import { MapSchema, Schema, type } from "@colyseus/schema";

console.log("COLYSEUS_PORT", process.env.COLYSEUS_PORT);
console.log("REDIS_HOST", process.env.REDIS_HOST);

const serverOptions = {
    presence: new RedisPresence({
        host: process.env.REDIS_HOST,
        port: process.env.REDIS_PORT ? Number(process.env.REDIS_PORT) : 6379,
    }),
};

class Player extends Schema {
    @type("string") sessionId: string = "";
    @type("number") x: number = 0;
    @type("number") y: number = 0;
}

class RoomState extends Schema {
    @type({ map: Player }) players = new MapSchema<Player>();
}

class TestRoom extends Room<RoomState> {
    maxClients = 4;
    onCreate(options: any) {
        console.log("TestRoom created!", options);

        this.setState(new RoomState());

        this.onMessage("move", this.onMove.bind(this));
    }
    onJoin(client: any, options: any) {
        const player = new Player();
        player.sessionId = client.sessionId;
        this.state.players.set(client.sessionId, player);
    }
    onMove(client: any, coors: {x: number, y: number}) {
        console.log("TestRoom received coords from", client.sessionId, ":", coors);
        const player = this.state.players.get(client.sessionId);
        if (player) {
            player.x = coors.x;
            player.y = coors.y;
        }
        // print out all players coords
        this.state.players.forEach((player) => {
            console.log(player.sessionId, player.x, player.y);
        });
    }
}

const gameServer = new Server(serverOptions);
gameServer.define("testroom", TestRoom);
gameServer.listen(port);

```

Server A Console Output
```bash
$ NODE_ENV=development COLYSEUS_PORT=2567 ts-node src/index.ts
COLYSEUS_PORT 2567
REDIS_HOST 127.0.0.1

       ___      _
      / __\___ | |_   _ ___  ___ _   _ ___
     / /  / _ \| | | | / __|/ _ \ | | / __|
    / /__| (_) | | |_| \__ \  __/ |_| \__ \
    \____/\___/|_|\__, |___/\___|\__,_|___/
                  |___/

Multiplayer Framework for Node.js · Open-source

💖 Sponsor Colyseus on GitHub → https://github.com/sponsors/endel
🌟 Give it a star on GitHub → https://github.com/colyseus/colyseus
☁️  Deploy and scale your project on Colyseus Cloud → https://cloud.colyseus.io

Error: seat reservation expired.
    at WebSocketServer.onConnection (C:\..\state-sync-test\server\node_modules\@colyseus\ws-transport\build\WebSocketTransport.js:102:15)
    at WebSocketServer.emit (node:events:513:28)
    at WebSocketServer.emit (node:domain:489:12)
    at WebSocketServer.completeUpgrade (C:\..\state-sync-test\server\node_modules\ws\lib\websocket-server.js:358:5)
    at WebSocketServer.handleUpgrade (C:\..\state-sync-test\server\node_modules\ws\lib\websocket-server.js:271:10)
    at Server.upgrade (C:\..\state-sync-test\server\node_modules\ws\lib\websocket-server.js:107:16)
    at Server.emit (node:events:513:28)
    at Server.emit (node:domain:489:12)
    at onParserExecuteCommon (node:_http_server:903:14)
    at onParserExecute (node:_http_server:797:3)
TestRoom created! {}
TestRoom received coords from KzVK_m1Ou : { x: 2, y: 3 }
KzVK_m1Ou 2 3
TestRoom received coords from KzVK_m1Ou : { x: 66, y: 66 }
KzVK_m1Ou 66 66
```
Server B Console Output
```bash
$ NODE_ENV=development COLYSEUS_PORT=2568 ts-node src/index.ts
COLYSEUS_PORT 2568
REDIS_HOST 127.0.0.1

       ___      _
      / __\___ | |_   _ ___  ___ _   _ ___
     / /  / _ \| | | | / __|/ _ \ | | / __|
    / /__| (_) | | |_| \__ \  __/ |_| \__ \
    \____/\___/|_|\__, |___/\___|\__,_|___/
                  |___/

Multiplayer Framework for Node.js · Open-source

💖 Sponsor Colyseus on GitHub → https://github.com/sponsors/endel
🌟 Give it a star on GitHub → https://github.com/colyseus/colyseus
☁️  Deploy and scale your project on Colyseus Cloud → https://cloud.colyseus.io

TestRoom created! {}
TestRoom received coords from Fmy5Mx9hT : { x: 4, y: 5 }
Fmy5Mx9hT 4 5
TestRoom received coords from Fmy5Mx9hT : { x: 6, y: 7 }
Fmy5Mx9hT 6 7
TestRoom received coords from Fmy5Mx9hT : { x: 77, y: 77 }
Fmy5Mx9hT 77 77
```

**Comment by endel** [comment]
Hi @bwasia, when using multiple processes you must also provide a `"publicAddress"` config option for the server: https://docs.colyseus.io/scalability/#recommendation-using-a-load-balancer

This is required so the client-side makes the request directly to the node that created the room and has the room in memory. Each room and each process is stateful in Colyseus. 

Hope this helps!

**Reply by bwasia** [reply]
hi @endel 

Thank you for your reply.

So, when you say "Each room and each process is stateful," you mean stateful for that node, right? Am I correct to assume this?

So, if a room mutates its state, will the nodes in the cluster be informed about the mutation only if the `publicAddress` option is set?

**Reply by eugene-egorov** [reply, marked as answer]
There is no any sharing state across processes. To sync some state across the rooms you can use redis

**Reply by bwasia** [reply]
Thank you!

**Comment by joshestein** [comment]
Hey @endel, thank you for a superb library!

I am wondering if there are any examples of syncing state across servers with Redis? I have read the documentation but am running into some difficulties getting things correctly setup.

**Reply by endel** [reply]
Hi @joshestein, may I ask about your use case?

**Reply by joshestein** [reply]
Hey 😄 

Sure. I am working on an interactive whiteboard, similar to Excalidraw. I want users to be able to collaborate (in realtime). I have successfully built a prototype of syncing state across clients within a single server. However, I would like to scale up the application across multiple servers.

I cannot guarantee that two clients collaborating on the same whiteboard will be in the same server. It seems that I will need redis to synchronise their state.

I have tried using a redis presence and making sure that every process has a `publicAddress`. However, I seem to always run into websocket errors and/or state not being propagated. I was wondering if there was an example you could point me to of setting up multiple servers with redis as an orchestrator.

**Reply by bwasia** [reply]
Hi @joshestein 

I solved the multi-server state synchronization by using Redis's pub/sub feature. Not sure if its the correct way, but it works for me until I find a better solution.

**Reply by endel** [reply]
> I cannot guarantee that two clients collaborating on the same whiteboard will be in the same server

I see, using version 0.15 and configuring Redis + `publicAddress` as mentioned in the docs would ensure Colyseus processes are aware of each other, and return the direct reference to the room your clients are connecting to: https://docs.colyseus.io/scalability/#step-2-making-your-colyseus-processes-accessible-to-the-internet

The key is that each process must be visible to the public internet via the provided `publicAddress` for each of them. e.g.:

**Using subdomains:**
- process1 -> https://colyseus-process-1.yourapp.io (`publicAddress: "colyseus-process-1.yourapp.io"`)
- process2 -> https://colyseus-process-2.yourapp.io  (`publicAddress: "colyseus-process-2.yourapp.io"`)

**(or) Using subroutes:**
- process1 -> https://backend.yourapp.io/colyseus-process-1 (`publicAddress: "backend.yourapp.io/colyseus-process-1"`)
- process2 -> https://backend.yourapp.io/colyseus-process-2 (`publicAddress: "backend.yourapp.io/colyseus-process-2"`)

You can configure NGINX to achieve either of the examples above!

**Reply by joshestein** [reply]
I appreciate the support! Thank you.

I have setup NGINX to use two subdomains. Then I am using pm2 to start two processes, setting each `publicAddress` to one of the corresponding domains.

I think the problem for me is getting the clients to connect. I've tried setting the Client's constructor endpoint to the public address port, a different port, the NGINX port and leaving it blank. I either run into a `Uncaught (in promise) ServerError: seat reservation expired` or a websocket connection failure.

I've also tried completely bypassing the Client and using the `matchmaker.presence` to directly pub/sub. However, I seem to be running into some bundling problems - this is something I am still trying to fix.

Again, I appreciate the help. I feel like I'm being dumb and missing something obvious to get the clients talking to each other, so your guidance is appreciated!

**Reply by zack1991** [reply]
Hi @endel ,

I also getting the same issue, I tried to configure the nginx but i am getting into error : **Error: seat reservation expired. at WebSocketServer.onConnection (/data/multiplayer-game-server-new/node_modules/@colyseus/ws-transport/build/WebSocketTransport.js:102:15)**

I am sharing my nginx configuration may you please suggest what i am doing wrong.

```
upstream sw-lb {
         server 127.0.0.1:1531;
        server 127.0.0.1:1532;
}

server {
	listen 80 default_server;
	listen [::]:80 default_server;

	root /var/www/html;

	# Add index.php to the list if you are using PHP
	index index.html index.htm index.nginx-debian.html;

	server_name _;
	
	location /sw-server-1/ {
		proxy_pass http://127.0.0.1:1531/;
                proxy_http_version 1.1;
                proxy_set_header Upgrade $http_upgrade;
                proxy_set_header Connection 'upgrade';
                proxy_set_header Host $host;
                proxy_set_header X-Real-IP $remote_addr;
                proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
                proxy_cache_bypass $http_upgrade;

	}
	
	location /sw-server-2/ {
		proxy_pass http://127.0.0.1:1532/;
                proxy_http_version 1.1;
                proxy_set_header Upgrade $http_upgrade;
                proxy_set_header Connection 'upgrade';
                proxy_set_header Host $host;
                proxy_set_header X-Real-IP $remote_addr;
                proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
                proxy_cache_bypass $http_upgrade;

	}
}

server {
	listen 8082;
	location / {
		proxy_pass http://sw-lb/; #upstream of nginx
                proxy_http_version 1.1;
                proxy_set_header Upgrade $http_upgrade;
                #proxy_set_header Connection 'upgrade';
                proxy_set_header Connection "";
				proxy_set_header Host $host;
                proxy_set_header X-Real-IP $remote_addr;
                proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
                proxy_cache_bypass $http_upgrade;
    }
}
```

If you don't mind, may you please share nginx configuration to achieve the example above you mentioned.

**Reply by endel** [reply]
Hi @zack1991, I don't see anything wrong with it. Feel free to send me an email at endel@colyseus.io if you'd like me to inspect your deployment. Hopefully we can identify what's wrong with it and share it here for everybody.

**Reply by zack1991** [reply]
Hi @endel , 
Thanks for your support, i will share the same with you on your email.

**Reply by joshestein** [reply]
Hi @endel,

Would you mind explaining what the client-side code should look like for load balancing with NGINX and using redis + `publicAddress`?

**Reply by endel** [reply]
Hi @joshestein, the client SDK detects if the server has sent its `publicAddress` and connects directly to it: https://github.com/colyseus/colyseus.js/blob/afb7642e73631855ed74d76de02f4e8d203a45c9/src/Client.ts#L208

So there's no special configuration in the client, just in the server. Feel free to send me an email too!

**Reply by zack1991** [reply]
Hi @endel ,

Thanks for your support my issue get resolved, the issue was related with package-lock.json. I deleted the package-lock.json and node_modules folder and then run **nmp install** after that i restarted my server and my issue get resolved.

**Reply by zack1991** [reply]
Hi @endel ,
My server is running perfectly when using client.js to connect to server. But when i am trying to join room with unity version 0.15.3 i am getting exception, But when running colyseus with single instance then it is working perfect. If you don't mind may you please help what is the problem, i will be thank full for that.

**Error Log :-**
Unable to connect to the remote server at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <9577ac7a62ef43179789031239ba8798>:0
at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in <9577ac7a62ef43179789031239ba8798>:0
at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in <9577ac7a62ef43179789031239ba8798>:0
at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in <9577ac7a62ef43179789031239ba8798>:0
at System.Runtime.CompilerServices.TaskAwaiter1[TResult].GetResult () [0x00000] in <9577ac7a62ef43179789031239ba8798>:0  at Colyseus.ColyseusClient+<ConsumeSeatReservation>d__251[T].MoveNext () [0x00200] in /Users/phonato-user/All Data/Bit Bucket/CamelCashCasino/Assets/Colyseus/Runtime/Colyseus/Models/ColyseusClient.cs:369

**Reply by endel** [reply]
Hi @zack1991 @VkDevPh, I've spent a few hours and couldn't reproduce this issue. (https://github.com/colyseus/colyseus-unity-sdk/issues/215). 

@VkDevPh also reported a `"error CS0173: Type of conditional expression cannot be determined because there is no implicit conversion between 'method group' and '<null>'"` ([screenshot](https://cdn.discordapp.com/attachments/526083188108296202/1121662228643459193/Screenshot_2023-06-23_at_10.04.13_AM.png)), which I also failed to reproduce.

Can you provide a reproduction scenario for these issues? (_Unfortunately, I feel like wasting time when trying to reproduce issues like this without the full context._)

**Reply by hamaddo** [reply]
Hi @endel,
Do I need to specify publicAddress if I have two docker containers with colyseus app isntance (3000, 3001 ports), I'm using nest js, so here is how I init server:

```ts
    const presence = new RedisPresence({
      host: redisUrlParts.host,
      port: redisUrlParts.port,
      password: redisPassword,
    });

    const driver = new RedisDriver({
      host: redisUrlParts.host,
      port: redisUrlParts.port,
      password: redisPassword,
    });

    const server = new Server({
      transport: new WebSocketTransport({
        server: app.getHttpServer(),
      }),
      presence: presence,
      driver: driver,
      logger: new Logger('Colyseus'),
    });
```

I'm also using nginx as load balancer, here is my config:

```nginx
upstream backend {
  ip_hash;
  server localhost:3000;
  server localhost:3001;
}

server {
    server_name your_domain.com;
    root /path/to/your/webroot;
    index index.html index.htm;

    location / {
        proxy_pass http://backend/;
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection 'upgrade';
        proxy_set_header Host $host;
        proxy_cache_bypass $http_upgrade;
        proxy_read_timeout 86400s;
        proxy_send_timeout 86400s;
        include proxy_params;
    }

    listen [::]:443 ssl; # managed by Certbot
    listen 443 ssl; # managed by Certbot
    ssl_certificate /path/to/your/fullchain.pem; # managed by Certbot
    ssl_certificate_key /path/to/your/privkey.pem; # managed by Certbot
    include /etc/letsencrypt/options-ssl-nginx.conf; # managed by Certbot
    ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem; # managed by Certbot
}

server {
    server_name your_domain.com;
    listen 80;
    listen [::]:80;

    location / {
        proxy_pass http://backend/;
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection 'upgrade';
        proxy_set_header Host $host;
        proxy_cache_bypass $http_upgrade;
        proxy_read_timeout 86400s;
        proxy_send_timeout 86400s;
        include proxy_params;
    }
}

```

but I'm still getting error:

```error
 ERROR [Colyseus] Error: seat reservation expired.
    at WebSocketServer.onConnection (/app/node_modules/@colyseus/ws-transport/build/WebSocketTransport.js:105:15)
    at WebSocketServer.emit (node:events:519:28)
    at WebSocketServer.completeUpgrade (/app/node_modules/@colyseus/ws-transport/node_modules/ws/lib/websocket-server.js:436:5)
    at WebSocketServer.handleUpgrade (/app/node_modules/@colyseus/ws-transport/node_modules/ws/lib/websocket-server.js:344:10)
    at Server.upgrade (/app/node_modules/@colyseus/ws-transport/node_modules/ws/lib/websocket-server.js:119:16)
    at Server.emit (node:events:519:28)
    at onParserExecuteCommon (node:_http_server:938:14)
    at onParserExecute (node:_http_server:839:3)

```

**Reply by endel** [reply]
Hi @hamaddo, the scalability docs have been improved on the upcoming version: https://deploy-preview-163--colyseus-docs.netlify.app/deployment/scalability

I would suggest enabling `DEBUG=colyseus:*` and checking if the traffic is coming as expected into your servers. 
[Colyseus Cloud](https://colyseus.io/cloud-managed-hosting/) have all these pieces sorted out, please consider subscribing if that works for you.
